### PR TITLE
[SPARK-42252][CORE] Add `spark.shuffle.localDisk.file.output.buffer` and deprecate `spark.shuffle.unsafe.file.output.buffer`

### PR DESCRIPTION
**Pull request by wayneguow**
### What changes were proposed in this pull request?
Deprecate spark.shuffle.unsafe.file.output.buffer and add a new config spark.shuffle.localDisk.file.output.buffer instead.

### Why are the changes needed?
The old config is desgined to be used in UnsafeShuffleWriter, but now it has been used in all local shuffle writers through LocalDiskShuffleMapOutputWriter, introduced by #25007. 

### Does this PR introduce _any_ user-facing change?
Old still works, advised to use new.

### How was this patch tested?
Passed existing tests.

### CODE REVIEW

**Comment by wayneguow** [comment]
cc @mccheah and @HyukjinKwon

**Review by dongjoon-hyun — DISMISSED** [review]
If you don't mind, let's target this proposal for Apache Spark 3.5.0.

**Comment by wayneguow** [comment]
> If you don't mind, let's target this proposal for Apache Spark 3.5.0.

Ok, it's good. I will make some changes.

**Comment by wayneguow** [comment]
Gentle ping @dongjoon-hyun, the target version of this proposal has been updated. Could you go on reviewing it when you have time? Thanks.

**Comment by dongjoon-hyun** [comment]
Thank you for your patience. Could you rebase this PR?

File: docs/core-migration-guide.md [hunk 1]
```diff
@@ -22,6 +22,10 @@ license: |
 * Table of contents
 {:toc}
 
+## Upgrading from Core 3.4 to 3.5
+
+- Since Spark 3.4, `spark.shuffle.unsafe.file.output.buffer` is deprecated though still works. Use `spark.shuffle.localDisk.file.output.buffer` instead.
```
Comment (dongjoon-hyun) [inline docs/core-migration-guide.md:27]: This should be `Since Spark 3.5`.

Comment (wayneguow) [reply docs/core-migration-guide.md]: I'm here, updated it, and waiting  for GA.

File: docs/core-migration-guide.md [hunk 2]
```diff
@@ -22,6 +22,10 @@ license: |
 * Table of contents
 {:toc}
 
+## Upgrading from Core 3.4 to 3.5
+
+- Since Spark 3.5, `spark.shuffle.unsafe.file.output.buffer` is deprecated though still works. Use `spark.shuffle.localDisk.file.output.buffer` instead.
```
[the remark this answers is not in the document]
Comment (wayneguow) [reply docs/core-migration-guide.md]: solved

**Comment by HyukjinKwon** [comment]
@wayneguow which PR made this change:

> The old config is desgined to be used in UnsafeShuffleWriter, but now it has been used in all local shuffle writers through LocalDiskShuffleMapOutputWriter.

? Would be easier for me to review if you can point it out.

**Comment by wayneguow** [comment]
> @wayneguow which PR made this change:
> 
> > The old config is desgined to be used in UnsafeShuffleWriter, but now it has been used in all local shuffle writers through LocalDiskShuffleMapOutputWriter.
> 
> ? Would be easier for me to review if you can point it out.

@HyukjinKwon The PR is #25007 .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side.

cc @HyukjinKwon , @cloud-fan , @Ngone51 , @mridulm

**Comment by dongjoon-hyun** [comment]
Too bad. Sorry but it seems that we forgot this PR, @wayneguow .

**Comment by dongjoon-hyun** [comment]
It seems too late because Apache Spark 3.5 is RC3 stage. If we want this still, we can do for Apache Spark 4.0.0.
- https://github.com/apache/spark/releases/tag/v3.5.0-rc3

File: docs/core-migration-guide.md [hunk 3]
```diff
@@ -48,7 +48,9 @@ license: |
 
 - Since Spark 4.0, the MDC (Mapped Diagnostic Context) key for Spark task names in Spark logs has been changed from `mdc.taskName` to `task_name`. To use the key `mdc.taskName`, you can set `spark.log.legacyTaskNameMdc.enabled` to `true`.
 
-- Since Spark 4.0, Spark performs speculative executions less agressively with `spark.speculation.multiplier=3` and `spark.speculation.quantile=0.9`. To restore the legacy behavior, you can set `spark.speculation.multiplier=1.5` and `spark.speculation.quantile=0.75`.
+- Since Spark 4.0, Spark performs speculative executions less aggressively with `spark.speculation.multiplier=3` and `spark.speculation.quantile=0.9`. To restore the legacy behavior, you can set `spark.speculation.multiplier=1.5` and `spark.speculation.quantile=0.75`.
```
Comment (wayneguow) [inline docs/core-migration-guide.md:51]: A minor typofix: `agressively` -> `aggressively `

**Comment by wayneguow** [comment]
Gentle ping @dongjoon-hyun , a pr from some time ago.

**Review by LuciferYang — APPROVED** [review]

**Comment by LuciferYang** [comment]
friendly ping @yaooqinn Do you have time to help take a look?

File: docs/core-migration-guide.md (hunk 3 above)
Comment (LuciferYang) [reply docs/core-migration-guide.md:51]: nit: I suggest fixing this typo in a separate minor pr

Comment (wayneguow) [reply docs/core-migration-guide.md]: @LuciferYang Sounds good, the separate mirror pr is #46864 .

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master, thank you @dongjoon-hyun @LuciferYang @wayneguow

**Pull request closed without merging**